tests/AnalyzeProperties.hs

```diff
@@ -0,0 +1,575 @@
+
+import           Pact.Eval                   (liftTerm, reduce)
+import           Pact.Native                 (lengthDef)
+import           Pact.Native.Ops
+import           Pact.Native.Time
+import           Pact.Repl                   (initPureEvalEnv)
+import           Pact.Typechecker            (typecheckTopLevel)
+import           Pact.Types.Exp              (Literal (..))
+import           Pact.Types.Native           (NativeDef)
+import           Pact.Types.Runtime          (PactError (..),
+                                              PactErrorType (EvalError),
+                                              runEval)
+import           Pact.Types.Term             (Term (TApp, TConst, TLiteral))
+import qualified Pact.Types.Term             as Pact
+import qualified Pact.Types.Type             as Pact
+import qualified Pact.Types.Typecheck        as Pact
+
+-- Note [EmptyInterval]: For both genDecimal and genInteger, it's possible that
+-- the range is too small to generate even one number. When this is true,
+-- these'll throw EmptyInterval. This is caught and discarded by the
+-- properties.
+
+genDecimal :: MonadGen m => NumSize -> m Decimal
+genDecimal size = do
+  places   <- Gen.word8 Range.constantBounded
+  mantissa <- genInteger (size * 10 ^ places)
+  pure $ mkDecimal $ Decimal.Decimal places mantissa
+
+genInteger :: MonadGen m => NumSize -> m Integer
+genInteger size = Gen.integral $
+  Range.exponentialFrom
+    (round   (midpoint size))
+    (ceiling (inf size))
+    (floor   (sup size))
+
+genArithOp :: MonadGen m => m ArithOp
+genArithOp = Gen.element [Add, Sub, Mul, Div] -- Pow, Log
+
+genUnaryArithOp :: MonadGen m => m UnaryArithOp
+genUnaryArithOp = Gen.element [Negate, Abs] -- Sqrt, Ln, Exp, Signum
+
+genRoundingLikeOp :: MonadGen m => m RoundingLikeOp
+genRoundingLikeOp = Gen.element [Round, Ceiling, Floor]
+
+genComparisonOp :: MonadGen m => m ComparisonOp
+genComparisonOp = Gen.element [Gt, Lt, Gte, Lte, Eq, Neq]
+
+type ECore = Existential (Core Analyze.Term)
+
+-- TODO: we might want to generalize this to a union of open intervals
+
+-- 'NumSize' bounds the values that a generated expression can evaluate to. We
+-- use this to avoid generating large numbers in computationally expensive
+-- spots.
+type NumSize = Interval Double
+
+data SizedType where
+  SizedInt     :: NumSize -> SizedType
+  SizedDecimal :: NumSize -> SizedType
+  SizedString  :: Int     -> SizedType
+  SizedTime    ::            SizedType
+  SizedBool    ::            SizedType
+  SizedKeySet  ::            SizedType
+  -- TODO: objects
+
+-- sizedLog :: NumSize -> (NumSize, NumSize)
+-- sizedLog size =
+--   let newExp  = logBase 10 size
+--       newBase = size ** (1 / newExp)
+--   in (newExp, newBase)
+
+arithSize :: ArithOp -> NumSize -> (NumSize, NumSize)
+arithSize op size = case op of
+  Add -> (size, size)
+  Sub -> (size, size)
+  Mul -> (sqrt size, sqrt size)
+  Div -> (size, size)
+  Pow -> undefined
```

Comment:
We don't generate any of these undefined cases because we don't symbolically interpret them (yet?).